src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java

```diff
@@ -0,0 +1,241 @@
+ * document name, the root directory needs a name. We call it "_dr_"; it could be anything, but it
+ * helps when viewing Azure in the console that it has an obvious name.
+ *
+ * <p>We don't store the root directory in the paths stored in file and directory entries. It is
+ * only used for the Azure lookup. When we refer to paths in the code we talk about: - lookup path -
+ * the path used for the Azure lookup. When building this path (and only this path) we prepended it
+ * with "_dr_" as the name for the root directory. - directory path - the directory path to the
+ * directory containing entry - not including the entry name - full path - the full path to the
+ * entry including the entry name.
+ *
+ * <p>Within the document we store the directory path to the entry and the entry name. That lets us
+ * use the indexes to find the entries in a directory.
+ *
+ * <p>It is an invariant that there are no empty directories. When a directory becomes empty on a
+ * delete, it is deleted. When a directory is needed, we create it. That is all done within
+ * transactions so there is never a time where the externally visible state violates that invariant.
+ */
+@Component
+public class TableDirectoryDao {
+  private final Logger logger = LoggerFactory.getLogger(FireStoreDirectoryDao.class);
+  private static final String TABLE_NAME = "dataset";
+  private static final String PARTITION_KEY = "partitionKey";
+  private final FileMetadataUtils fileMetadataUtils;
+
+  @Autowired
+  public TableDirectoryDao(FileMetadataUtils fileMetadataUtils) {
+    this.fileMetadataUtils = fileMetadataUtils;
+  }
+
+  // Note that this does not test for duplicates. If invoked on an existing path it will overwrite
+  // the entry. Existence checking is handled at upper layers.
+  public void createDirectoryEntry(
+      TableServiceClient tableServiceClient, FireStoreDirectoryEntry createEntry) {
+
+    tableServiceClient.createTableIfNotExists(TABLE_NAME);
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+
+    List<TableTransactionAction> createList = new ArrayList<>();
+
+    // Walk up the lookup directory path, finding missing directories we get to an
+    // existing one
+    // We will create the ROOT_DIR_NAME directory here if it does not exist.
+    String lookupDirPath = fileMetadataUtils.makeLookupPath(createEntry.getPath());
+    for (String testPath = lookupDirPath;
+        !testPath.isEmpty();
+        testPath = fileMetadataUtils.getDirectoryPath(testPath)) {
+
+      // !!! In this case we are using a lookup path
+      if (lookupByFilePath(tableServiceClient, testPath) != null) {
+        break;
+      }
+
+      FireStoreDirectoryEntry dirToCreate = fileMetadataUtils.makeDirectoryEntry(testPath);
+      String rowKey = fileMetadataUtils.encodePathAsFirestoreDocumentName(testPath);
+      TableEntity entity =
+          FireStoreDirectoryEntry.toTableEntity(PARTITION_KEY, rowKey, dirToCreate);
+      TableTransactionAction t =
```

Comment:
No sure if there's value in using a transaction for a single entity/row. Furthermore, given the fixed partition key, this transaction will lock the entire table.

Reply:
We create a variable number of entities here so I put it into a single transaction. Specifically, we create an entity for the `createEntity` (corresponding to the file that is being ingested), and for any directories in the file path that do not have a record. 

Also that's a good point about this locking the entire table. What is the behavior for a simultaneous request that tries to read/write to the table? Does it retry depending on the client configuration you mentioned above?

Reply:
Okay, with a variable number of entities makes sense. Transactions are scoped at the partition level. So if you have a single partition, you will lock the whole table and serialize the write operations, with the potential impact of affecting the throughput of the system.  Using a correct partition scheme should help with this as well.

Reply:
These entities all have different partition keys now, so I removed the transaction here